keel-core/src/main/kotlin/com/netflix/spinnaker/keel/services/ApplicationService.kt

```diff
@@ -85,16 +89,30 @@ class ApplicationService(
   fun pin(application: String, pin: EnvironmentArtifactPin, user: String) {
     val config = repository.getDeliveryConfigForApplication(application)
     repository.pinEnvironment(config, pin.copy(pinnedBy = user))
+    publisher.publishEvent(
```

Comment:
You could argue this is redundant with the record in the `environment_artifact_pin` table, but this provides an easy way to follow the trail of breadcrumbs about pinning and other events from a single source (`event` table). It also has a counterpart for unpinning below.

Reply:
Will open a separate PR for events.